File: ProjectEuler/051-100/S51.cs
using System.Collections.Generic;
using System.Linq;

namespace ProjectEuler
{
    public class S51 : ISolution
    {
        private static readonly string[] patterns = new string[] { "110001", "101001", "100101",
            "100011", "011001", "010101", "010011", "001101", "001011", "000111" };

        public string GetAnswer()
        {
            var groups = new List<List<int>>();
            for (int i = 0; i < patterns.Length; i++)
            {
                groups.Add(new List<int>());
            }

            var primes = Utils.GenPrimes(1000000).Where(l => l > 100000).Select(l => (int)l).ToArray();
            foreach (var p in primes)
            {
                var pStr = p.ToString();
                for (int i = 0; i < patterns.Length; i++)
                {
                    string pattern = patterns[i];
                    char[] chars = new char[3];
                    int index = 0;
                    for (int j = 0; j < pattern.Length; j++)
                    {
                        if (pattern[j] == '0')
                        {
                            chars[index] = pStr[j];
                            index++;
                        }
                    }

                    if (chars[0] == chars[1] && chars[1] == chars[2])
                    {
                        groups[i].Add(p);
                    }
                }
            }

            for (int i = 0; i < patterns.Length; i++)
            {
                var group = groups[i];
                var eight = HandleGroup(group, patterns[i]);
                if (eight.Count > 0)
                {
                    return eight.First().First().ToString();
                }
            }

            return string.Empty;
        }

        private static List<List<int>> HandleGroup(List<int> primes, string pattern)
        {
            Dictionary<int, List<int>> subGroups = new Dictionary<int, List<int>>();
            for (int i = 0; i < primes.Count; i++)
            {
                int p = primes[i];
                int key = 0;
                for (int j = pattern.Length - 1; j >= 0; j--)
                {
                    if (pattern[j] == '1')
                    {
                        key *= 10;
                        key += p % 10;
                    }

                    p /= 10;
                }

                if (subGroups.ContainsKey(key))
                {
                    subGroups[key].Add(primes[i]);
                }
                else
                {
                    subGroups[key] = new List<int> { primes[i] };
                }
            }

            return subGroups.Values.Where(ints => ints.Count == 8).ToList();
        }
    }
}

File: src/051-100/S51.cc
#include "S51.h"
#include "../Utils.h"
#include <map>
#include <string>
#include <vector>

using namespace std;

namespace ProjectEuler
{
std::string S51::GetAnswer()
{
	std::vector<std::vector<int>> groups(patterns.size());

	auto all_primes = Utils::GenPrimes(1000000);
	std::vector<int> primes;
	for (long long p : all_primes)
	{
		if (p > 100000)
		{
			primes.push_back((int)p);
		}
	}

	for (int p : primes)
	{
		std::string pStr = std::to_string(p);
		for (size_t i = 0; i < patterns.size(); i++)
		{
			std::string pattern = patterns[i];
			std::vector<char> extracted;
			for (size_t j = 0; j < pattern.size(); j++)
			{
				if (pattern[j] == '0')
				{
					extracted.push_back(pStr[j]);
				}
			}

			// We expect 3 replaced digits for 6-digit number?
			// Problem says "part of the number". Patterns in S51.h seem to have three 0s.
			// e.g. "110001".
			// Check if all extracted digits are same
			if (extracted.size() == 3 && extracted[0] == extracted[1] && extracted[1] == extracted[2])
			{
				groups[i].push_back(p);
			}
		}
	}

	for (size_t i = 0; i < patterns.size(); i++)
	{
		auto group = groups[i];
		auto eight = HandleGroup(group, patterns[i]);
		if (eight.size() > 0)
		{
			// Return smallest element of the first family found
			// Primes are sorted by generation order (increasing).
			return std::to_string(eight[0][0]);
		}
	}

	return "";
}

std::vector<std::vector<int>> S51::HandleGroup(std::vector<int> primes, std::string pattern)
{
	std::map<int, std::vector<int>> subGroups;
	for (int p : primes)
	{
		// Construct key from '1' positions
		// We need to match digits at '1' positions.
		// Can construct an integer key or string key.
		// Original logic: iterate backwards, build int.
		// "110001" -> key is digits at 0,1,5.

		std::string s = std::to_string(p);
		int key = 0;
		for (size_t j = 0; j < pattern.size(); j++)
		{
			if (pattern[j] == '1')
			{
				key = key * 10 + (s[j] - '0');
			}
		}

		subGroups[key].push_back(p);
	}

	std::vector<std::vector<int>> result;
	for (const auto &[key, family] : subGroups)
	{
		if (family.size() == 8)
		{
			result.push_back(family);
		}
	}
	return result;
}

} // namespace ProjectEuler
